src/test/scala/org/broadinstitute/dsde/workbench/leonardo/api/ProxyRoutesSpec.scala

```diff
@@ -0,0 +1,186 @@
+      handled shouldBe false
+    }
+    Get("/foo").addHeader(LocalHostHeader) ~> proxyRoutes ~> check {
+      handled shouldBe false
+    }
+    Get("/").addHeader(LocalHostHeader) ~> proxyRoutes ~> check {
+      handled shouldBe false
+    }
+  }
+
+  it should "pass through paths" in {
+    Get("/notebooks/test").addHeader(LocalHostHeader) ~> proxyRoutes ~> check {
+      responseAs[Data].path shouldEqual "/notebooks/test"
+    }
+    Get("/notebooks/test2").addHeader(LocalHostHeader) ~> proxyRoutes ~> check {
+      responseAs[Data].path shouldEqual "/notebooks/test2"
+    }
+    Get("/notebooks/test3").addHeader(LocalHostHeader) ~> proxyRoutes ~> check {
+      responseAs[Data].path shouldEqual "/notebooks/test3"
+    }
+  }
+
+  it should "pass through query string params" in {
+    Get("/notebooks/test").addHeader(LocalHostHeader) ~> proxyRoutes ~> check {
+      responseAs[Data].qs shouldBe None
+    }
+    Get("/notebooks/test?foo=bar&baz=biz").addHeader(LocalHostHeader) ~> proxyRoutes ~> check {
+      responseAs[Data].qs shouldEqual Some("foo=bar&baz=biz")
+    }
+  }
+
+  it should "pass through http methods" in {
+    Get("/notebooks/test").addHeader(LocalHostHeader) ~> proxyRoutes ~> check {
+      responseAs[Data].method shouldBe "GET"
+    }
+    Post("/notebooks/test").addHeader(LocalHostHeader) ~> proxyRoutes ~> check {
+      responseAs[Data].method shouldBe "POST"
+    }
+    Put("/notebooks/test").addHeader(LocalHostHeader) ~> proxyRoutes ~> check {
+      responseAs[Data].method shouldBe "PUT"
+    }
+  }
+
+  it should "pass through headers" in {
+    Get("/notebooks/test")
+      .addHeader(LocalHostHeader)
+      .addHeader(RawHeader("foo", "bar"))
+      .addHeader(RawHeader("baz", "biz")) ~> proxyRoutes ~> check {
+      responseAs[Data].headers should contain allElementsOf Map("foo" -> "bar", "baz" -> "biz")
+    }
+  }
+
+  /**
+    * Akka-http TestKit doesn't seem to support websocket servers.
+    * So for websocket tests, manually create a server binding on port 9000 for the proxy.
+    */
+  def withWebsocketProxy[T](testCode: => T): T = {
+    val bindingFuture = Http().bindAndHandle(proxyRoutes, "0.0.0.0", 9000)
+    try {
+      testCode
+    } finally {
+      bindingFuture.flatMap(_.unbind())
+    }
+  }
+
+  it should "proxy websockets" in withWebsocketProxy {
+    // Sink for incoming data from the WebSocket
+    val incoming = Sink.head[String]
+
+    // Source outgoing data over the WebSocket
+    val outgoing = Source.single(TextMessage("Leonardo"))
+
+    // Flow to hit the proxy server
+    val webSocketFlow = Http().webSocketClientFlow(WebSocketRequest(Uri("ws://localhost:9000/notebooks/websocket"))).map {
+      case m: TextMessage.Strict => m.text
+      case _ => throw new IllegalArgumentException("ProxyRoutesSpec only supports strict messages")
+    }
+
+    // Glue together the source, sink, and flow. Returns a tuple of:
```

Comment:
I will try to add more explanation to `ProxyRoutes` and `ProxyRoutesSpec` (assuming I understand it myself). This akka-stream stuff is hard to follow.